### ISSUE TITLE
[API Proposal]: Custom schema generation in STJ

### ISSUE THREAD

**Issue by tkvalvik** [dotnet/runtime#108778]
### Background and motivation

We already have custom converters, but not custom schemas to accompany them. The schema and conversion are semantically tightly coupled, and it would be nice to be able to define both in the manner.

Types with custom converters may not comply to STJ-conventions, and may not even be well behaved, for instance as described in https://github.com/dotnet/runtime/issues/108755. 

Custom schema generation could be implemented similarly to how converters are, with direct equivalents for `JsonConverter<T>`, configuration on `JsonSerializerOptions` and the `JsonConverterAttribute`.

### API Proposal

```csharp
// Base class or interface for generators
public abstract class JsonSchemaGenerator<T>
{
    public abstract JsonSchema GenerateSchema(JsonSchemaGeneratorContext context);
}

// List of SchemaGenerators on JsonSerializerOptions
public class JsonSerializerOptions
{
    public IList<JsonSchemaGenerator> SchemaGenerators { get; }

    public void AddSchemaGenerator(JsonSerializerOptions options);
}

// Attribute for properties, same as JsonConverterAttribute
public class JsonSchemaAttribute : JsonAttribute {
    public Type? GeneratorType { get; private set; }
}
```

### API Usage

```csharp

public class MyCustomSchemaGenerator: JsonSchemaGenerator<MyDto>
{
    public override JsonSchema GenerateSchema(JsonSchemaGeneratorContext context)
    {
        var schema = new JsonSchema() {....}
        // Build schema
        return schema;
    }
}

// Apply to options
var options = new JsonSerializerOptions();
options.AddSchemaGenerator(new MyCustomSchemaGenerator());

// Or, apply to dto:
[JsonSchema(typeof(MyCustomSchemaGenerator))]
public class MyDto {...]
```
Microsoft.AspNetCore.OpenApi

### Risks

- Feature bloat.
- Possible partial overlap with Microsoft.AspNetCore.OpenApi?

**Comment by gregsdennis** [comment]
There are quite a few steps that need to happen before getting here, but it's a good idea. .Net 9 is introducing a JSON Schema exporter to support some OpenAPI functionality, but it's only generating schemas using the nodes model. This gave a quick win for the needed OpenAPI support without committing to a full schema model, which will take a considerable amount of time to design.

In the meantime, you should have a look at my JsonSchema.Net, which provides some [validation-during-serialization support](https://docs.json-everything.net/schema/serialization/). The serialization piece is not AOT compatible, but the rest of the library is.

There's also a [schema generation extension](https://docs.json-everything.net/schema/schemagen/schema-generation/) for the lib that's a lot more customizable than the coming schema exporter.

**Comment by eiriktsarpalis** [comment]
> There's also a [schema generation extension](https://docs.json-everything.net/schema/schemagen/schema-generation/) for the lib that's a lot more customizable than the coming schema exporter.

IIRC that component isn't currently AOT compatible correct? Is it conceivable that you can use the same contract metadata APIs that `JsonSchemaExporter` reads from to make it work everywhere?

**Comment by eiriktsarpalis** [comment]
Closing as duplicate of #105769

**Comment by gregsdennis** [comment]
> IIRC that component isn't currently AOT compatible correct?

That's correct. It uses reflection. 

> Is it conceivable that you can use the same contract metadata APIs that JsonSchemaExporter reads from to make it work everywhere?

That would require people who want to generate schemas also have a serialization context with all of the types on it, which isn't currently required.  I can explore it, but I think it would result in a bunch of duplicated code.
